### PR TITLE
Made `std::hint::unreachable_unchecked` a const fn.

### PR DESCRIPTION
**Pull request by q2p**
r? @oli-obk
tracking issue: #53188

### CODE REVIEW

**Comment by jonas-schievink** [comment]
You probably also want to add a test that executes the intrinsic to initialize a `const` item

**Comment by Centril** [comment]
Including a test case where we trigger UB would also be good I think.

cc @rust-lang/wg-const-eval

**Comment by lzutao** [comment]
What's the use case for constifying this function?
Does never type signal the compiler early and making constifying needlessly?

**Comment by RalfJung** [comment]
Yeah, you'll need to actually implement the intrinsic as raising an [unreachable error](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/mir/interpret/enum.UndefinedBehaviorInfo.html#variant.Unreachable).

Turns out we [forgot about that in Miri](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a7817b0a5c88049a70d9731d1c082c8c). ;)

**Comment by q2p** [comment]
> Including a test case where we trigger UB would also be good I think.

I don't know how to write a test for an undefined behavior. Should function be called when initialising a const value, or at runtime? How should assertion or stderr output look like?

> Yeah, and for that to work you'll need to actually implement the intrinsic as raising an [unreachable error](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/mir/interpret/enum.UndefinedBehaviorInfo.html#variant.Unreachable).

Sorry, but I don't know how to do that.

File: src/test/ui/consts/const_unsafe_unreachable.rs [hunk 1]
```diff
@@ -1,18 +1,28 @@
...
         42 => 34,
         _ => std::hint::unreachable_unchecked(),
     }
 }
 
-const BAR:u8 = unsafe { foo(42) };
+const unsafe fn f_std_intrinsics_unreachable(x: u8) -> u8 {
+    match x {
+        17 => 22,
+        _ => std::intrinsics::unreachable(),
```
Comment (RalfJung) [inline src/test/ui/consts/const_unsafe_unreachable.rs]: `std::hint::unreachable_unchecked` is a thing wrapper around the intrinsic, I do not see the point of also testing the intrinsic directly? I don't think we usually do that.

The test @eddyb and me asked for was a case where `unreachable_unchecked` is used *incorrectly*, where UB is triggered by actually reaching that function. Like, `f_std_hint_unreachable(23)`.

**Comment by RalfJung** [comment]
> Sorry, but I don't know how to do that.

You need to add a new arm in [this match](https://github.com/rust-lang/rust/blob/8549cfed4bbcdd28ef3a36e5de72c000e32f650e/src/librustc_mir/interpret/intrinsics.rs#L89) on `sym::unreachable` that then does `throw_ub!(Unreachable)`.

**Comment by q2p** [comment]
I have added an arm to the match inside of `librustc_mir/interpret/intrinsics.rs` as RalfJung asked.

```
...
sym::unreachable_code => throw_ub!(Unreachable),
...
```

But I have doubts that it is a right way, because everything else in the file calls `throw_ub_format!` macro and not `throw_ub!`.

Also this change, didn't have any effect on output when running the UB test.

---

When I tried to implement a test where function calls `unreachable_unchecked()`, it prints out:
```
...
error: any use of this value will cause an error
  --> /media/p2a/rust/src/libcore/hint.rs:52:5
   |
LL |     intrinsics::unreachable()
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     "calling intrinsic `unreachable`" needs an rfc before being allowed inside constants
   |     inside call to `std::hint::unreachable_unchecked` at /media/p2a/rust/src/test/ui/consts/const_unsafe_unreachable_ub.rs:8:14
...
```
Which references `libcore/hint.rs:52:5` and says that `"calling intrinsic unreachable" needs an rfc`.
So I'm not sure if this is a desired error message or how to normalize this stderr.

Clearly I can't make this PR, because I have no idea what kind of checks / code should be implemented for `intrinsics::unreachable()` to have a meaningful error message on incorrect use, or how to prevent the incorrect use.

And I don't want to endlessly keep bothering people with questions, so I'm gonna close this PR.

**Comment by RalfJung** [comment]
But you're almost there. ;)

> But I have doubts that it is a right way, because everything else in the file calls throw_ub_format! macro and not throw_ub!.

Yeah that's okay. We have specific error constants for common errors, everything else uses the `_format` variant. Since we have a variant for "entered unreachable code", that is okay to use.

But your mistake is that you used `unreachable_code`, when the intrinsic is called `unreachable`. That also explains why you are getting the wrong error message.

If `sym::unreachable` does not exist, you have to add it here:

https://github.com/rust-lang/rust/blob/1add455ec6f81045e7651c6225902823f5d4fbfa/src/librustc_span/symbol.rs#L783

**Pull request closed without merging**